databend-jdbc/src/main/java/com/databend/jdbc/DatabendConnection.java

```diff
@@ -635,16 +675,35 @@ DatabendClient startQueryWithFailover(String sql, StageAttachment attach) throws
                 throw new SQLException("Error start query: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
             }
             try {
+                // route hint is used when transaction occured,
+                // transaction procedure:
+                // 1. server return session body where txn state is active
+                // 2. when there is a active transaction, it will route all query to target route hint uri if exists
+                // 3. if there is not active transaction, it will use load balancing policy to choose a host to execute query
+                String query_id = UUID.randomUUID().toString();
+                String candidateHost = this.driverUri.getUri(query_id).toString();
                 if (!inActiveTransaction()) {
-                    this.routeHint = randRouteHint();
+                    this.routeHint = uriRouteHint(candidateHost);
                 }
+                // checkout the host to use from route hint
+                if (this.routeHint != null && !this.routeHint.isEmpty()) {
+                    URI uri = parseRouteHint(this.routeHint);
```

Comment:
add a comment here? there're also a case like multi-cluster warehouse, the client will get the route hint returned from the server, and in this case, this route hint is not parseable.

it's important that `parseRouteHint` can not raise an exception when it met unparseable routehint

Reply:
public static URI parseRouteHint(String routeHint) {
        if (routeHint == null || routeHint.isEmpty()) {
            return null;
        }
        URI target;
        try {
            if (routeHint.charAt(routeHint.length() - 1) != SPECIAL_CHAR) {
                return null;
            }
            // Remove the special character
            String encodedUri = routeHint.substring(0, routeHint.length() - 1);

            // Decode the Base64 string
            byte[] decodedBytes = Base64.getDecoder().decode(encodedUri);
            String decodedUri = new String(decodedBytes);

            return create(decodedUri);
        } catch (Exception e) {
            logger.log(Level.FINE, "Failed to parse route hint: " + routeHint, e);
            return null;
        }
    }
    
    it catch all exception and will return null if exception occur

Reply:
great, i guess it's not a bad thing to add a comment here in case we forgot the case of the server returned route hint.

Reply:
already added. PTAL.

Reply:
like "// besides the generated route hint, we may also got an route hint from the server side, which is unparseable. parseRouteHint will try to retrieve the target embedded in the route hint, and will return null on unparseable route hint, NOT raising any exception."